irohad/ordering/impl/on_demand_ordering_gate.cpp

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ordering/impl/on_demand_ordering_gate.hpp"
+
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+using namespace iroha;
+using namespace iroha::ordering;
+
+namespace {
+  /**
+   * Shortcut for calculating next reject round
+   * @param round current round
+   * @return next reject round
+   */
+  constexpr transport::RoundType nextRejectRound(
+      const transport::RoundType &round) {
+    return {round.first, round.second + 1};
+  }
+}  // namespace
+
+OnDemandOrderingGate::OnDemandOrderingGate(
+    std::shared_ptr<OnDemandOrderingService> ordering_service,
+    std::shared_ptr<transport::OdOsNotification> network_client,
+    std::shared_ptr<network::ProposalGate> proposal_gate,
+    std::function<shared_model::interface::TransactionBatch(
+        std::shared_ptr<const shared_model::interface::Transaction>)>
+        batch_factory,
+    rxcpp::observable<BlockRoundEventType> events,
+    transport::RoundType initial_round)
+    : ordering_service_(std::move(ordering_service)),
+      network_client_(std::move(network_client)),
+      proposal_gate_(std::move(proposal_gate)),
+      batch_factory_(std::move(batch_factory)),
+      subscription_(events.subscribe([this](auto event) {
+        // exclusive lock
+        std::lock_guard<std::shared_timed_mutex> lock(mutex_);
+
+        visit_in_place(event,
+                       [this](BlockEvent block) {
+                         // block committed, increment block round
+                         current_round_ = {block.height, 1};
+                       },
+                       [this](EmptyEvent empty) {
+                         current_round_ = nextRejectRound(current_round_);
+                       });
+        this->vote();
+      })),
+      current_round_(initial_round) {}
+
+void OnDemandOrderingGate::propagateTransaction(
+    std::shared_ptr<const shared_model::interface::Transaction> transaction)
+    const {
+  propagateBatch(batch_factory_(std::move(transaction)));
```

Comment:
I suggest not doing support this method because we have been moving to batch API.
So, maybe throw an exception instead?